### PR TITLE
fix paste crash when section is invalid

### LINKED ISSUE #78219: [macOS/Windows] TextField exception when pasting via ctrl-V after clearing the field
## Steps to Reproduce

1. Run `flutter create bug`.
2. Add a TextField with an explicit TextEditingController
3. Set onEditingComplete to clear the controller's text
4. Press enter, then ctrl-V a piece of text from your clipboard

**Expected results:**

Text from clipboard pastes into the empty text field.

**Actual results:**

selection.textBefore(text) from editable.dart produces an exception and the clipboard data is not pasted:

> [ERROR:flutter/lib/ui/ui_dart_state.cc(186)] Unhandled Exception: RangeError: Value not in range: -1
> #0      _StringBase.substring (dart:core-patch/string_patch.dart:399:7)
> #1      TextRange.textBefore (dart:ui/text.dart:1533:17)
> #2      RenderEditable._handleShortcuts
> package:flutter/…/rendering/editable.dart:822
> <asynchronous suspension>

**Solved by:**

Adding a check before the value is assigned a new TextEditingValue:
```
var newString = data.text!;
if (text.isNotEmpty) {
          newString = selection.textBefore(text) +
              data.text! +
              selection.textAfter(text);
}
```

<details>
  <summary>Logs</summary>

Sample app:

```
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  TextEditingController _controller;

  @override
  void initState() {
    super.initState();
    _controller = TextEditingController();
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: TextField(
          controller: _controller,
          onEditingComplete: () => _controller.text = '',
        ),
      ),
    );
  }
}
```

```
flutter analyze
Analyzing testapp...                                                    
No issues found! (ran in 2.0s)
```

```
[√] Flutter (Channel stable, 2.0.1, on Microsoft Windows [Version 10.0.19041.867], locale en-GB)
    • Flutter version 2.0.1 at C:\Work\flutter
    • Framework revision c5a4b4029c (11 days ago), 2021-03-04 09:47:48 -0800
    • Engine revision 40441def69
    • Dart version 2.12.0

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:\Users\Sean\AppData\Local\Android\Sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_SDK_ROOT = C:\Users\Sean\AppData\Local\Android\Sdk
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.8.5)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.8.31005.135
    • Windows 10 SDK version 10.0.18362.0

[√] Android Studio (version 4.1.0)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] VS Code (version 1.54.1)
    • VS Code at C:\Users\Sean\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.20.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.19041.867]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 89.0.4389.82
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 88.0.705.74

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by xu-baolin**
Fixes #78219 

This patch makes `_handleShortcuts` stronger when processing `ctrl + v` event.

### CODE REVIEW

File: packages/flutter/test/material/text_field_test.dart [hunk 1]
```diff
@@ -4624,6 +4624,57 @@ void main() {
     expect(find.text(expected), findsOneWidget, reason: 'Because text contains ${controller.text}');
   });
 
+  // Regressing test for https://github.com/flutter/flutter/issues/78219
+  testWidgets('Paste do not crash when the section is inValid', (WidgetTester tester) async {
```
Comment (justinmc) [inline packages/flutter/test/material/text_field_test.dart:4628]:
```diff
- 'Paste do not crash when the section is inValid'
+ 'Paste does not crash when the selection is invalid'
```

**Review by justinmc — APPROVED** [review]
LGTM, thanks for fixing this!

The ability of selection to be (-1,-1) keeps causing problems.  CC @LongCatIsLooong and @chunhtai in case you guys want to take a look, since we were recently talking about similar stuff.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 2]
```diff
@@ -1782,9 +1782,9 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
       final ClipboardData? data = await Clipboard.getData(Clipboard.kTextPlain);
       if (data != null) {
         value = TextEditingValue(
-          text: selection.textBefore(text) + data.text! + selection.textAfter(text),
```
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/editable.dart:1785]: for selection is invalid we will append it to the end, I think this is probably fine, but do we know what the native behavior is?

**Review by chunhtai — APPROVED** [review]
LGTM

**Review by LongCatIsLooong — COMMENTED** [review]
I feel we shouldn't do anything when there's no selection. WDYT?

I'm not sure what the author of the issue meant by "Set onEditingComplete to clear the controller's text". If they are referring to `clear` doesn't it now set the selection to (0, 0)? If they just forgot to set the selection then it's a programmer error (but crashing is still bad).

File: packages/flutter/lib/src/rendering/editable.dart (hunk 2 above)
Comment (xu-baolin) [reply packages/flutter/lib/src/rendering/editable.dart:1785]: I am not sure we can set the selection invalid on the native platform cc @LongCatIsLooong @justinmc

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/editable.dart:1785]: Yes (at least for iOS and android): 
- iOS https://developer.apple.com/documentation/uikit/uitextinput/1614541-selectedtextrange can be `nil`
- Android there can be no span at all, and you can remove the selection from your text using https://developer.android.com/reference/android/text/Selection?hl=en#removeSelection(android.text.Spannable)

**Comment by LongCatIsLooong** [comment]
@justinmc yes. The developer can manually set the selection to `(-1, -1)`, the `TextEditingController.text` setter sets it to `(-1, -1)`, and I believe some text formatters do that as well. 

Currently `RenderEditable` still shows caret when the selection is `(-1, -1)` but I think that's a bug (currently working to fix that).

**Comment by justinmc** [comment]
Got it, sounds good.  Should this PR wait until you're done?

**Comment by LongCatIsLooong** [comment]
@justinmc I don't think this depends on https://github.com/flutter/flutter/issues/76573. I brought that up because it's a bit relevant.

**Comment by justinmc** [comment]
Sounds good 👍

**Comment by xu-baolin** [comment]
@LongCatIsLooong hey, If I understand correctly when selection is invalid, pasting doesn't do anything, right?

**Comment by LongCatIsLooong** [comment]
@xu-baolin yeah I would think "don't do anything if there's no selection" makes sense, it would feel a bit random to pick an insertion point for the user when there's no selection.

**Comment by LongCatIsLooong** [comment]
Actually android's `replaceText` treats `(-1, -1)` as `(0, 0)`: https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/view/inputmethod/BaseInputConnection.java;l=808-813;drc=master;bpv=1;bpt=0

But looking at the commit https://cs.android.com/android/_/android/platform/frameworks/base/+/a465a170ce5d7155580fd308d1e50092365117e4 and the commit message, I'm not sure what the rationale is. I'll take a look at the bugs listed in the commit message and see if I can find anything there.

**Comment by LongCatIsLooong** [comment]
The 3 mentioned bugs no longer exist.

**Comment by xu-baolin** [comment]
> The 3 mentioned bugs no longer exist.

The bugs were more than ten years ago. What is your opinion now?

**Comment by LongCatIsLooong** [comment]
since I couldn't find out why Android handles it that way, I'm still leaning towards not doing anything

**Review by LongCatIsLooong — APPROVED** [review]

**Pull request merged**